frame/safe-mode/src/lib.rs

```diff
@@ -0,0 +1,362 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2022 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use frame_support::{
+	pallet_prelude::*,
+	traits::{
+		CallMetadata, Contains, Currency, Defensive, GetCallMetadata, PalletInfoAccess,
+		ReservableCurrency,
+	},
+};
+use frame_system::pallet_prelude::*;
+use sp_runtime::traits::Saturating;
+use sp_std::{convert::TryInto, prelude::*};
+
+mod benchmarking;
+mod mock;
+mod tests;
+
+pub use pallet::*;
+
+type BalanceOf<T> =
+	<<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(PhantomData<T>);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The overarching event type.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// Contains all calls that can be dispatched when the safe-mode is enabled.
+		///
+		/// The `SafeMode` pallet is always included and does not need to be added here.
+		type SafeModeFilter: Contains<Self::Call>;
+
+		/// Currency type for this pallet.
+		type Currency: ReservableCurrency<Self::AccountId>;
+
+		/// How long the safe-mode will stay active when enabled with [`Pallet::enable`].
+		#[pallet::constant]
+		type EnableDuration: Get<Self::BlockNumber>;
```

Comment:
shouldn't this be configurable based on the deposit? Or perhaps this can be the granularity, and you can enable it 5x if you pay 5x deposit.

Reply:
Yes it will. As Gav proposed it will be configurable per Origin implementation.  
So that there is something like:  
```rust
let (duration, staking_amount, who) = ensure_signed(origin)?;
```

Reply:
Okay for the first iteration I have it configured per permissioned origin. So the `ForceEnableOrigin` now looks like this `EnsureOrigin<Self::Origin, Success = Self::BlockNumber>`.  
This allows for granularity for permissioned origins. For the permissionless enabling I would like to keep it simple for a first version.